kolibri/core/content/views.py

```diff
@@ -140,23 +135,28 @@ def parse_html(content):
                 )
             },
         )
-
         # Currently, html5lib strips the doctype, but it's important for correct rendering, so check the original
         # content for the doctype and, if found, prepend it to the content serialized by html5lib
         doctype = None
         try:
-            encoding = parser.documentEncoding
-            if not encoding:
-                encoding = "utf-8"
-            doctype = re.match(
-                "(?i)<!DOCTYPE[^<>]*(?:<!ENTITY[^<>]*>[^<>]*)?>",
-                content.decode(encoding),
+            # Now parse the content as a dom tree instead, so that we capture
+            # any doctype node as a dom node that we can read.
+            tree_builder_dom = html5lib.treebuilders.getTreeBuilder("dom")
+            parser_dom = html5lib.HTMLParser(
+                tree_builder_dom, namespaceHTMLElements=False
             )
-        except Exception as e:
-            # Losing the doctype could lead to some rendering issues, but they are usually not severe enough
-            # to be worth stopping the content from loading entirely.
-            logger.warning("Attempt to determine doctype failed while parsing HTML.")
-            logger.warning("Error details: {}".format(e))
+            tree = parser_dom.parse(content)
+            # By HTML Spec if doctype is included, it must be the first thing
+            # in the document, so it has to be the first child node of the document
+            doctype_node = tree.childNodes[0]
+
+            # Check that this node is in fact a doctype node
+            if doctype_node.nodeType == doctype_node.DOCUMENT_TYPE_NODE:
+                # render to a string by calling the toxml method
+                # toxml uses single quotes by default, replace with ""
+                doctype = doctype_node.toxml().replace("'", '"')
+        except (html5lib.html5parser.ParseError, IndexError):
```

Comment:
It's worth noting that this will introduce a regression if we get an error other than these two errors, as it will not only fail to get the doctype, but the failure will result in a 500 and not render anything at all, instead of the quirky HTML that it serves today.

Reply:
I am struggling to understand what other errors you think might be thrown - even the parser error catch is a bit of overkill, as that error would already have been thrown by the outer parse invocation if there was something fundamentally broken.

As the parser has already run by this point without triggering an exception, the only place an error could occur is `doctype_node = tree.childNodes[0]` if the tree happens to be completely empty. Even this is very unlikely as passing any string to `parser_dom.parse` always produces at least one child node.

Reply:
If there is one thing that life and my many years of coding has taught me, it's that I can't anticipate every possible problem that could happen, and also that code changes over time and so code that may not have thrown a certain error before may start throwing it later (or may start to error out with the introduction of new content unlike our current content base).

A lot of times, throwing a hard error is the best you can do, because you would need to know the details of the error in order to know what to do about it. In this case, though, we do know that the page will render the page without the doctype, and in the case of most of our existing HTML5 apps, we know they will also work. So what is the argument against allowing them to continue loading?